opencart 1.5.6.4/upload/catalog/controller/smailyforopencart/cron_customers.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/**
+ * Samaly subscribers synchronization.
+ */
+class ControllerSmailyForOpencartCronCustomers extends Controller {
+
+    public function index() {
+        // Customer model.
+        $this->load->model('account/customer');
+        // Load Smaily settings.
+        $this->load->model('setting/setting');
+        // Load Smaily helper.
+        $this->load->model('smailyforopencart/helper');
+        
+        $settings = $this->model_setting_setting->getSetting('smaily')['smaily_customer_sync'];
+
+        // Validate cron token.
+        if (!array_key_exists('token', $settings) ||
+            empty($this->request->get['token']) ||
+            $settings['token'] !== $this->request->get['token']
+        ) {
+            die('Unauthorized');
+        }
+        if ($settings['enabled'] == 1) {
+
+            // Get unsubscribers from smaily.
+            $unsubscribers_filter = array(
+                'list' => 2,
+            );
+
+            $unsubscribers = $this->model_smailyforopencart_helper->apiCall('contact', $unsubscribers_filter);
```

Comment:
Unsubscribing customers in this manner has some issues:
a) by default, `/api/contact.php?list=2` request will return only the first 25000 unsubscribers;
b) fetching 25000 subscribers at once can be taxing (and can cause "Out of memory" exception) on some systems when Smaily database contains many custom data fields;
c) unsubscribing customers from Opencart is slow when customers are updated one-by-one;

You can use this pattern to resolve those issues:
```php
$offset = 0;
while (true) {
    $unsubscribers = $this->model_smailyforopencart_helper->apiCall('contact', [
        'list' => 2,
        'offset' => $offset,
        'limit' => 2500,
    ]);

    // Collect unsubscribe emails.

    // unsubscribeCustomers method would compile a single update query.
    $this->model_smailyforopencart_helper->unsubscribeCustomers($list_of_emails_to_unsubscribe);

    $offset += 1;
}
```